File: python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_wallet_provider.py
"""CDP wallet provider."""

import json
import os
from decimal import Decimal
from typing import Any

from cdp import (
    Cdp,
    ExternalAddress,
    MnemonicSeedPhrase,
    Wallet,
    WalletData,
    hash_message,
    hash_typed_data_message,
)
from eth_account.typed_transactions import DynamicFeeTransaction
from pydantic import BaseModel, Field
from web3 import Web3
from web3.types import BlockIdentifier, ChecksumAddress, HexStr, TxParams

from ..network import Network
from .evm_wallet_provider import EvmWalletProvider


class CdpProviderConfig(BaseModel):
    """Configuration options for CDP providers."""

    api_key_name: str | None = Field(None, description="The CDP API key name")
    api_key_private_key: str | None = Field(None, description="The CDP API private key")


class CdpWalletProviderConfig(CdpProviderConfig):
    """Configuration options for CDP wallet provider."""

    chain_id: int | None = Field(None, description="The chain id")
    network_id: str | None = Field("base-sepolia", description="The network id")
    mnemonic_phrase: str | None = Field(None, description="The mnemonic phrase of the wallet")
    rpc_url: str | None = Field(None, description="The RPC URL")
    wallet_data: str | None = Field(None, description="The data of the CDP Wallet as a JSON string")


class CdpWalletProvider(EvmWalletProvider):
    """A wallet provider that uses the CDP SDK."""

    def __init__(self, config: CdpWalletProviderConfig | None = None):
        """Initialize CDP wallet provider.

        Args:
            config: Configuration options for the CDP provider.

        """
        if not config:
            config = CdpWalletProviderConfig()

        try:
            api_key_name = config.api_key_name or os.getenv("CDP_API_KEY_NAME")
            api_key_private_key = config.api_key_private_key or os.getenv("CDP_API_KEY_PRIVATE_KEY")

            if api_key_name and api_key_private_key:
                Cdp.configure(
                    api_key_name=api_key_name,
                    private_key=api_key_private_key.replace("\\n", "\n"),
                )
            else:
                Cdp.configure_from_json()

            network_id = config.network_id or os.getenv("NETWORK_ID", "base-sepolia")
            chain_id = config.chain_id or os.getenv("CHAIN_ID", "84532")
            rpc_url = config.rpc_url or os.getenv("RPC_URL", "https://sepolia.base.org")

            if not network_id:
                raise ValueError("NETWORK_ID is required")

            if not chain_id:
                raise ValueError("CHAIN_ID is required")

            if not rpc_url:
                raise ValueError("RPC_URL is required")

            if config.wallet_data:
                wallet_data = WalletData.from_dict(json.loads(config.wallet_data))
                self._wallet = Wallet.import_data(wallet_data)
            elif config.mnemonic_phrase:
                phrase = MnemonicSeedPhrase(config.mnemonic_phrase)
                self._wallet = Wallet.import_wallet(phrase, network_id)
            else:
                self._wallet = Wallet.create(network_id=network_id)

            self._address = self._wallet.default_address.address_id
            self._network = Network(
                protocol_family="evm",
                network_id=network_id,
                chain_id=chain_id,
            )
            self._web3 = Web3(Web3.HTTPProvider(rpc_url))

        except ImportError as e:
            raise ImportError(
                "Failed to import cdp. Please install it with 'pip install cdp-sdk'."
            ) from e
        except Exception as e:
            raise ValueError(f"Failed to initialize CDP wallet: {e!s}") from e

    def get_address(self) -> str:
        """Get the wallet address."""
        return self._address

    def get_balance(self) -> Decimal:
        """Get the wallet balance in native currency."""
        if not self._wallet:
            raise Exception("Wallet not initialized")

        balance = self._wallet.balance("eth")
        return Decimal(str(Web3.to_wei(balance, "ether")))

    def get_name(self) -> str:
        """Get the name of the wallet provider."""
        return "cdp_wallet_provider"

    def get_network(self) -> Network:
        """Get the current network."""
        return self._network

    def native_transfer(self, to: str, value: Decimal) -> str:
        """Transfer the native asset of the network.

        Args:
            to: The destination address
            value: The amount to transfer in whole units (e.g. '1.5' for 1.5 ETH)

        Returns:
            The transaction hash as a string

        """
        if not self._wallet:
            raise Exception("Wallet not initialized")

        try:
            transfer_result = self._wallet.transfer(
                amount=value,
                asset_id="eth",
                destination=Web3.to_checksum_address(to),
                gasless=False,
            )

            transfer_result.wait()
            tx_hash = transfer_result.transaction_hash

            if not tx_hash:
                raise Exception("Transaction hash not found")

            return tx_hash
        except Exception as e:
            raise Exception(f"Failed to transfer native tokens: {e!s}") from e

    def read_contract(
        self,
        contract_address: ChecksumAddress,
        abi: list[dict[str, Any]],
        function_name: str,
        args: list[Any] | None = None,
        block_identifier: BlockIdentifier = "latest",
    ) -> Any:
        """Read data from a smart contract."""
        contract = self._web3.eth.contract(address=contract_address, abi=abi)
        func = contract.functions[function_name]
        if args is None:
            args = []
        return func(*args).call(block_identifier=block_identifier)

    def sign_message(self, message: str | bytes) -> HexStr:
        """Sign a message using the wallet's private key."""
        if not self._wallet:
            raise Exception("Wallet not initialized")

        message_hash = hash_message(message)
        payload_signature = self._wallet.sign_payload(message_hash)

        return payload_signature.signature

    def sign_typed_data(self, typed_data: dict[str, Any]) -> HexStr:
        """Sign typed data according to EIP-712 standard."""
        if not self._wallet:
            raise Exception("Wallet not initialized")

        typed_data_message_hash = hash_typed_data_message(typed_data)

        payload_signature = self._wallet.sign_payload(typed_data_message_hash)

        return payload_signature.signature

    def sign_transaction(self, transaction: TxParams) -> HexStr:
        """Sign an EVM transaction."""
        if not self._wallet:
            raise Exception("Wallet not initialized")

        dynamic_fee_tx = DynamicFeeTransaction.from_dict(transaction)

        tx_hash_bytes = dynamic_fee_tx.hash()
        tx_hash_hex = tx_hash_bytes.hex()

        payload_signature = self._wallet.sign_payload(tx_hash_hex)
        return payload_signature.signature

    def send_transaction(self, transaction: TxParams) -> HexStr:
        """Send a signed transaction to the network."""
        self._prepare_transaction(transaction)

        signature = self.sign_transaction(transaction)

        transaction["r"] = int(signature[2:66], 16)
        transaction["s"] = int(signature[66:130], 16)
        transaction["v"] = int(signature[130:132], 16) - 27

        signed_dynamic_fee_tx = DynamicFeeTransaction.from_dict(transaction)

        signed_bytes = signed_dynamic_fee_tx.payload()

        external_address = ExternalAddress(
            self._wallet.network_id, self._wallet.default_address.address_id
        )
        broadcasted_transaction = external_address.broadcast_external_transaction(
            "02" + signed_bytes.hex()
        )

        return broadcasted_transaction.transaction_hash

    def get_transaction_receipt(self, tx_hash: HexStr) -> dict[str, Any]:
        """Get the transaction receipt for a given transaction hash."""
        raise NotImplementedError

    def wait_for_transaction_receipt(
        self, tx_hash: HexStr, timeout: float = 120, poll_latency: float = 0.1
    ) -> dict[str, Any]:
        """Wait for transaction confirmation and return receipt."""
        return self._web3.eth.wait_for_transaction_receipt(
            tx_hash, timeout=timeout, poll_latency=poll_latency
        )

    def _prepare_transaction(self, transaction: TxParams) -> TxParams:
        """Prepare EIP-1559 transaction for signing."""
        if transaction["to"]:
            transaction["to"] = Web3.to_bytes(hexstr=transaction["to"])
        else:
            transaction["to"] = b""

        transaction["from"] = self._address
        transaction["value"] = int(transaction.get("value", 0))
        transaction["type"] = 2
        transaction["chainId"] = self._network.chain_id

        nonce = self._web3.eth.get_transaction_count(self._address)
        transaction["nonce"] = nonce

        data_field = transaction.get("data", b"")
        if isinstance(data_field, str) and data_field.startswith("0x"):
            data_bytes = bytes.fromhex(data_field[2:])

        transaction["data"] = data_bytes

        max_priority_fee_per_gas, max_fee_per_gas = self._estimate_fees()
        transaction["maxPriorityFeePerGas"] = max_priority_fee_per_gas
        transaction["maxFeePerGas"] = max_fee_per_gas

        gas = self._web3.eth.estimate_gas(transaction)
        transaction["gas"] = gas

        del transaction["from"]

        return transaction

    def _estimate_fees(self, multiplier=1.2):
        """Estimate fees."""

        def get_base_fee():
            latest_block = self._web3.eth.get_block("latest")
            base_fee = latest_block["baseFeePerGas"]
            # Multiply by 1.2 to give some buffer
            return int(base_fee * multiplier)

        base_fee_per_gas = get_base_fee()
        max_priority_fee_per_gas = Web3.to_wei(0.1, "gwei")
        max_fee_per_gas = base_fee_per_gas + max_priority_fee_per_gas

        return (max_priority_fee_per_gas, max_fee_per_gas)

    def export_wallet(self) -> WalletData:
        """Export the wallet.

        Returns:
            The wallet data.

        """
        if not self._wallet:
            raise Exception("Wallet not initialized")

        return self._wallet.export_data()

    def deploy_contract(
        self,
        solidity_version: str,
        solidity_input_json: str,
        contract_name: str,
        constructor_args: dict[str, Any],
    ) -> Any:
        """Deploy a smart contract.

        Args:
            solidity_version: The version of the Solidity compiler to use
            solidity_input_json: The JSON input for the Solidity compiler
            contract_name: The name of the contract to deploy
            constructor_args: Key-value map of constructor args

        Returns:
            The deployed contract instance

        Raises:
            Exception: If wallet is not initialized or deployment fails

        """
        if not self._wallet:
            raise Exception("Wallet not initialized")

        try:
            return self._wallet.deploy_contract(
                solidity_version=solidity_version,
                solidity_input_json=solidity_input_json,
                contract_name=contract_name,
                constructor_args=constructor_args,
            )
        except Exception as e:
            raise Exception(f"Failed to deploy contract: {e!s}") from e

    def deploy_nft(self, name: str, symbol: str, base_uri: str) -> Any:
        """Deploy a new NFT (ERC-721) smart contract.

        Args:
            name: The name of the collection
            symbol: The token symbol for the collection
            base_uri: The base URI for token metadata

        Returns:
            The deployed NFT contract instance

        Raises:
            Exception: If wallet is not initialized or deployment fails

        """
        if not self._wallet:
            raise Exception("Wallet not initialized")

        try:
            return self._wallet.deploy_nft(
                name=name,
                symbol=symbol,
                base_uri=base_uri,
            )
        except Exception as e:
            raise Exception(f"Failed to deploy NFT: {e!s}") from e

    def deploy_token(self, name: str, symbol: str, total_supply: str) -> Any:
        """Deploy an ERC20 token contract.

        Args:
            name: The name of the token
            symbol: The symbol of the token
            total_supply: The total supply of the token

        Returns:
            The deployed token contract instance

        Raises:
            Exception: If wallet is not initialized or deployment fails

        """
        if not self._wallet:
            raise Exception("Wallet not initialized")

        try:
            return self._wallet.deploy_token(
                name=name,
                symbol=symbol,
                total_supply=total_supply,
            )
        except Exception as e:
            raise Exception(f"Failed to deploy token: {e!s}") from e

    def trade(self, amount: str, from_asset_id: str, to_asset_id: str) -> str:
        """Trade a specified amount of a from asset to a to asset for the wallet.

        Args:
            amount (str): The amount of the from asset to trade, e.g. `15`, `0.000001`.
            from_asset_id (str): The from asset ID to trade (e.g., "eth", "usdc", or a valid contract address).
            to_asset_id (str): The to asset ID to trade (e.g., "eth", "usdc", or a valid contract address).

        Returns:
            str: A message containing the trade details.

        Raises:
            Exception: If the trade fails or if the wallet is not initialized.

        """
        if not self._wallet:
            raise Exception("Wallet not initialized")

        try:
            trade_result = self._wallet.trade(
                amount=amount,
                from_asset_id=from_asset_id,
                to_asset_id=to_asset_id,
            ).wait()

            return "\n".join(
                [
                    f"Traded {amount} of {from_asset_id} for {trade_result.to_amount} of {to_asset_id}.",
                    f"Transaction hash for the trade: {trade_result.transaction.transaction_hash}",
                    f"Transaction link for the trade: {trade_result.transaction.transaction_link}",
                ]
            )
        except Exception as e:
            raise Exception(f"Error trading assets: {e!s}") from e

File: typescript/agentkit/src/wallet-providers/cdpWalletProvider.ts
import { version } from "../../package.json";
import { Decimal } from "decimal.js";
import {
  createPublicClient,
  serializeTransaction,
  TransactionRequest,
  TransactionSerializable,
  http,
  keccak256,
  PublicClient,
  TransactionReceipt,
  Hex,
  Address,
  Chain,
} from "viem";
import { EvmWalletProvider } from "./evmWalletProvider";
import { EvmNetwork } from "../network";
import {
  Coinbase,
  CreateERC20Options,
  CreateTradeOptions,
  ExternalAddress,
  SmartContract,
  Trade,
  Wallet,
  WalletData,
  hashTypedDataMessage,
  hashMessage,
  TypedDataDomain,
  TypedDataField,
} from "@coinbase/coinbase-sdk";
import { NETWORK_ID, NETWORK_ID_TO_CHAIN_ID, NETWORK_ID_TO_VIEM_CHAIN } from "../network/network";

/**
 * Configuration options for the CDP Providers.
 */
export interface CdpProviderConfig {
  /**
   * The CDP API Key Name.
   */
  apiKeyName?: string;

  /**
   * The CDP API Key Private Key.
   */
  apiKeyPrivateKey?: string;
}

/**
 * Configuration options for the CdpActionProvider.
 */
export interface CdpWalletProviderConfig extends CdpProviderConfig {
  /**
   * The CDP Wallet.
   */
  wallet?: Wallet;

  /**
   * The address of the wallet.
   */
  address?: string;

  /**
   * The network of the wallet.
   */
  network?: EvmNetwork;
}

/**
 * Configuration options for the CDP Agentkit with a Wallet.
 */
interface ConfigureCdpAgentkitWithWalletOptions extends CdpWalletProviderConfig {
  /**
   * The data of the CDP Wallet as a JSON string.
   */
  cdpWalletData?: string;

  /**
   * The mnemonic phrase of the wallet.
   */
  mnemonicPhrase?: string;
}

/**
 * A wallet provider that uses the Coinbase SDK.
 */
export class CdpWalletProvider extends EvmWalletProvider {
  /**
   * Reads a contract.
   *
   * @param params - The parameters to read the contract.
   * @returns The response from the contract.
   *
   * @description Since the readContract is a complex generic function, it is much simpler to just copy the function signature and assign it in the constructor.
   */
  readContract: PublicClient["readContract"];

  #cdpWallet?: Wallet;
  #address?: string;
  #network?: EvmNetwork;
  #publicClient: PublicClient;

  /**
   * Constructs a new CdpWalletProvider.
   *
   * @param config - The configuration options for the CdpWalletProvider.
   */
  private constructor(config: CdpWalletProviderConfig) {
    super();

    this.#cdpWallet = config.wallet;
    this.#address = config.address;
    this.#network = config.network;
    this.#publicClient = createPublicClient({
      chain: NETWORK_ID_TO_VIEM_CHAIN[config.network!.networkId!] as Chain,
      transport: http(),
    });

    this.readContract = this.#publicClient.readContract;
  }

  /**
   * Configures a new CdpWalletProvider with a wallet.
   *
   * @param config - Optional configuration parameters
   * @returns A Promise that resolves to a new CdpWalletProvider instance
   * @throws Error if required environment variables are missing or wallet initialization fails
   */
  public static async configureWithWallet(
    config: ConfigureCdpAgentkitWithWalletOptions = {},
  ): Promise<CdpWalletProvider> {
    if (config.apiKeyName && config.apiKeyPrivateKey) {
      Coinbase.configure({
        apiKeyName: config.apiKeyName,
        privateKey: config.apiKeyPrivateKey,
        source: "agentkit",
        sourceVersion: version,
      });
    } else {
      Coinbase.configureFromJson();
    }

    let wallet: Wallet;

    const mnemonicPhrase = config.mnemonicPhrase || process.env.MNEMONIC_PHRASE;
    const networkId = (config.network?.networkId ||
      process.env.NETWORK_ID ||
      Coinbase.networks.BaseSepolia) as NETWORK_ID;

    try {
      if (config.wallet) {
        wallet = config.wallet;
      } else if (config.cdpWalletData) {
        const walletData = JSON.parse(config.cdpWalletData) as WalletData;
        wallet = await Wallet.import(walletData);
      } else if (mnemonicPhrase) {
        wallet = await Wallet.import({ mnemonicPhrase: mnemonicPhrase }, networkId);
      } else {
        wallet = await Wallet.create({ networkId: networkId });
      }
    } catch (error) {
      throw new Error(`Failed to initialize wallet: ${error}`);
    }

    const address = (await wallet.getDefaultAddress())?.getId();

    const network = {
      protocolFamily: "evm" as const,
      chainId: NETWORK_ID_TO_CHAIN_ID[networkId],
      networkId: networkId,
    };

    const cdpWalletProvider = new CdpWalletProvider({
      wallet,
      address,
      network,
    });

    return cdpWalletProvider;
  }

  /**
   * Signs a message.
   *
   * @param message - The message to sign.
   * @returns The signed message.
   */
  async signMessage(message: string): Promise<Hex> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    const messageHash = hashMessage(message);
    const payload = await this.#cdpWallet.createPayloadSignature(messageHash);

    if (payload.getStatus() === "pending" && payload?.wait) {
      await payload.wait(); // needed for Server-Signers
    }

    return payload.getSignature() as Hex;
  }

  /**
   * Signs a typed data object.
   *
   * @param typedData - The typed data object to sign.
   * @param typedData.domain - The domain parameters for the EIP-712 message, including the name, version, chainId, and verifying contract.
   * @param typedData.types - The types definitions for the EIP-712 message, represented as a record of type names to their fields.
   * @param typedData.message - The actual data object to hash, conforming to the types defined.
   *
   * @returns The signed typed data object.
   */
  async signTypedData(typedData: {
    domain: TypedDataDomain;
    types: Record<string, Array<TypedDataField>>;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    message: Record<string, any>;
  }): Promise<Hex> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    const messageHash = hashTypedDataMessage(
      typedData.domain!,
      typedData.types!,
      typedData.message!,
    );

    const payload = await this.#cdpWallet.createPayloadSignature(messageHash);

    if (payload.getStatus() === "pending" && payload?.wait) {
      await payload.wait(); // needed for Server-Signers
    }

    return payload.getSignature() as Hex;
  }

  /**
   * Signs a transaction.
   *
   * @param transaction - The transaction to sign.
   * @returns The signed transaction.
   */
  async signTransaction(transaction: TransactionRequest): Promise<Hex> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    const serializedTx = serializeTransaction(transaction as TransactionSerializable);
    const transactionHash = keccak256(serializedTx);

    const payload = await this.#cdpWallet.createPayloadSignature(transactionHash);

    if (payload.getStatus() === "pending" && payload?.wait) {
      await payload.wait(); // needed for Server-Signers
    }

    return payload.getSignature() as Hex;
  }

  /**
   * Sends a transaction.
   *
   * @param transaction - The transaction to send.
   * @returns The hash of the transaction.
   */
  async sendTransaction(transaction: TransactionRequest): Promise<Hex> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    const preparedTransaction = await this.prepareTransaction(
      transaction.to!,
      transaction.value!,
      transaction.data!,
    );

    const signature = await this.signTransaction({
      ...preparedTransaction,
    } as TransactionRequest);

    const signedPayload = await this.addSignatureAndSerialize(preparedTransaction, signature);

    const externalAddress = new ExternalAddress(this.#cdpWallet.getNetworkId(), this.#address!);

    const tx = await externalAddress.broadcastExternalTransaction(signedPayload.slice(2));

    return tx.transactionHash as Hex;
  }

  /**
   * Prepares a transaction.
   *
   * @param to - The address to send the transaction to.
   * @param value - The value of the transaction.
   * @param data - The data of the transaction.
   * @returns The prepared transaction.
   */
  async prepareTransaction(
    to: Address,
    value: bigint,
    data: Hex,
  ): Promise<TransactionSerializable> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    const nonce = await this.#publicClient!.getTransactionCount({
      address: this.#address! as Address,
    });

    const feeData = await this.#publicClient!.estimateFeesPerGas();

    const gas = await this.#publicClient!.estimateGas({
      account: this.#address! as Address,
      to,
      value,
      data,
      maxFeePerGas: feeData.maxFeePerGas,
      maxPriorityFeePerGas: feeData.maxPriorityFeePerGas,
    });

    const chainId = this.#network!.chainId;

    return {
      to,
      value,
      data,
      nonce,
      maxFeePerGas: feeData.maxFeePerGas,
      maxPriorityFeePerGas: feeData.maxPriorityFeePerGas,
      gas,
      chainId,
      type: "eip1559",
    };
  }

  /**
   * Adds signature to a transaction and serializes it for broadcast.
   *
   * @param transaction - The transaction to sign.
   * @param signature - The signature to add to the transaction.
   * @returns A serialized transaction.
   */
  async addSignatureAndSerialize(
    transaction: TransactionSerializable,
    signature: Hex,
  ): Promise<Hex> {
    // Decode the signature into its components
    const r = `0x${signature.slice(2, 66)}` as const; // First 32 bytes
    const s = `0x${signature.slice(66, 130)}` as const; // Next 32 bytes
    const v = BigInt(parseInt(signature.slice(130, 132), 16)); // Last byte

    return serializeTransaction(transaction, { r, s, v });
  }

  /**
   * Gets the address of the wallet.
   *
   * @returns The address of the wallet.
   */
  getAddress(): Address {
    if (!this.#address) {
      throw new Error("Address not initialized");
    }

    return this.#address as Address;
  }

  /**
   * Gets the network of the wallet.
   *
   * @returns The network of the wallet.
   */
  getNetwork(): EvmNetwork {
    if (!this.#network) {
      throw new Error("Network not initialized");
    }

    return this.#network;
  }

  /**
   * Gets the name of the wallet provider.
   *
   * @returns The name of the wallet provider.
   */
  getName(): string {
    return "cdp_wallet_provider";
  }

  /**
   * Gets the balance of the wallet.
   *
   * @returns The balance of the wallet in wei
   */
  async getBalance(): Promise<bigint> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    const balance = await this.#cdpWallet.getBalance("eth");
    return BigInt(balance.mul(10 ** 18).toString());
  }

  /**
   * Waits for a transaction receipt.
   *
   * @param txHash - The hash of the transaction to wait for.
   * @returns The transaction receipt.
   */
  async waitForTransactionReceipt(txHash: Hex): Promise<TransactionReceipt> {
    return await this.#publicClient!.waitForTransactionReceipt({ hash: txHash });
  }

  /**
   * Creates a trade.
   *
   * @param options - The options for the trade.
   * @returns The trade.
   */
  async createTrade(options: CreateTradeOptions): Promise<Trade> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    return this.#cdpWallet.createTrade(options);
  }

  /**
   * Deploys a token.
   *
   * @param options - The options for the token deployment.
   * @returns The deployed token.
   */
  async deployToken(options: CreateERC20Options): Promise<SmartContract> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    return this.#cdpWallet.deployToken(options);
  }

  /**
   * Deploys a contract.
   *
   * @param options - The options for contract deployment
   * @param options.solidityVersion - The version of the Solidity compiler to use (e.g. "0.8.0+commit.c7dfd78e")
   * @param options.solidityInputJson - The JSON input for the Solidity compiler containing contract source and settings
   * @param options.contractName - The name of the contract to deploy
   * @param options.constructorArgs - Key-value map of constructor args
   *
   * @returns A Promise that resolves to the deployed contract instance
   * @throws Error if wallet is not initialized
   */
  async deployContract(options: {
    solidityVersion: string;
    solidityInputJson: string;
    contractName: string;
    constructorArgs: Record<string, unknown>;
  }): Promise<SmartContract> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    return this.#cdpWallet.deployContract(options);
  }

  /**
   * Deploys a new NFT (ERC-721) smart contract.
   *
   * @param options - Configuration options for the NFT contract deployment
   * @param options.name - The name of the collection
   * @param options.symbol - The token symbol for the collection
   * @param options.baseURI - The base URI for token metadata.
   *
   * @returns A Promise that resolves to the deployed SmartContract instance
   * @throws Error if the wallet is not properly initialized
   * @throws Error if the deployment fails for any reason (network issues, insufficient funds, etc.)
   */
  async deployNFT(options: {
    name: string;
    symbol: string;
    baseURI: string;
  }): Promise<SmartContract> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    return this.#cdpWallet.deployNFT(options);
  }

  /**
   * Transfer the native asset of the network.
   *
   * @param to - The destination address.
   * @param value - The amount to transfer in Wei.
   * @returns The transaction hash.
   */
  async nativeTransfer(to: Address, value: string): Promise<Hex> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    const transferResult = await this.#cdpWallet.createTransfer({
      amount: new Decimal(value),
      assetId: Coinbase.assets.Eth,
      destination: to,
      gasless: false,
    });

    const result = await transferResult.wait();

    if (!result.getTransactionHash()) {
      throw new Error("Transaction hash not found");
    }

    return result.getTransactionHash() as Hex;
  }

  /**
   * Exports the wallet.
   *
   * @returns The wallet's data.
   */
  async exportWallet(): Promise<WalletData> {
    if (!this.#cdpWallet) {
      throw new Error("Wallet not initialized");
    }

    return this.#cdpWallet.export();
  }
}
